src/catalog/catalog_set.cpp

```diff
@@ -1,86 +1,199 @@
+CatalogEntry* CatalogSet::traverseVersionChainsForTransaction(Transaction* transaction,
+    CatalogEntry* currentEntry) const {
+    while (currentEntry) {
+        if (currentEntry->getTimestamp() == transaction->getID()) {
+            // This entry is created by the current transaction.
+            break;
+        }
+        if (currentEntry->getTimestamp() <= transaction->getStartTS()) {
+            // This entry was committed before the current transaction starts.
+            break;
+        }
+        currentEntry = currentEntry->getPrev();
+    }
+    return currentEntry;
+}
+
+CatalogEntry* CatalogSet::getCommittedEntry(CatalogEntry* entry) const {
+    while (entry) {
+        if (entry->getTimestamp() < Transaction::START_TRANSACTION_ID) {
+            break;
+        }
+        entry = entry->getPrev();
+    }
+    return entry;
+}
+
+bool CatalogSet::checkWWConflict(Transaction* transaction, CatalogEntry* entry) const {
+    return (entry->getTimestamp() >= Transaction::START_TRANSACTION_ID &&
+               entry->getTimestamp() != transaction->getID()) ||
+           (entry->getTimestamp() < Transaction::START_TRANSACTION_ID &&
+               entry->getTimestamp() > transaction->getStartTS());
+}
+
+void CatalogSet::dropEntry(Transaction* transaction, const std::string& name) {
+    KU_ASSERT(containsEntry(transaction, name));
+    // TODO: Should handle rel group and rdf graph.
+    auto tombstone = createDummyEntry(name);
+    tombstone->setTimestamp(transaction->getID());
+    auto tombstonePtr = tombstone.get();
+    emplace(std::move(tombstone));
+    if (transaction->getStartTS() > 0) {
+        KU_ASSERT(transaction->getID() != 0);
+        transaction->addCatalogEntry(this, tombstonePtr->getPrev());
+    }
+}
+
+void CatalogSet::alterEntry(Transaction* transaction, const binder::BoundAlterInfo& alterInfo) {
+    KU_ASSERT(containsEntry(transaction, alterInfo.tableName));
+    auto entry = getEntry(transaction, alterInfo.tableName);
+    KU_ASSERT(entry->getType() == CatalogEntryType::NODE_TABLE_ENTRY ||
+              entry->getType() == CatalogEntryType::REL_TABLE_ENTRY ||
+              entry->getType() == CatalogEntryType::REL_GROUP_ENTRY ||
+              entry->getType() == CatalogEntryType::RDF_GRAPH_ENTRY);
+    auto tableEntry = ku_dynamic_cast<CatalogEntry*, TableCatalogEntry*>(entry);
+    auto newEntry = tableEntry->alter(alterInfo);
+    newEntry->setTimestamp(transaction->getID());
+    if (alterInfo.alterType == AlterType::RENAME_TABLE) {
+        // We treat rename table as drop and create.
+        dropEntry(transaction, alterInfo.tableName);
+        createEntry(transaction, std::move(newEntry));
+        return;
+    }
+    emplace(std::move(newEntry));
+    if (transaction->getStartTS() > 0) {
+        KU_ASSERT(transaction->getID() != 0);
+        transaction->addCatalogEntry(this, entry);
+    }
+}
+
+case_insensitive_map_t<CatalogEntry*> CatalogSet::getEntries(Transaction* transaction) {
```

Comment:
can we give this type an alias:
e.g. `using CatalogEntrySet case_insensitive_map_t<CatalogEntry*>`